pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/ControllerBasedBrokerSelector.java

```diff
@@ -0,0 +1,140 @@
+      long controllerFetchRetriesIntervalMills, long controllerFetchScheduleIntervalMills) {
+    this.controllerUrls = controllerUrl.split(",");
+    this.controllerFetchRetries = controllerFetchRetries;
+    this.controllerFetchRetriesIntervalMills = controllerFetchRetriesIntervalMills;
+    this.controllerFetchScheduleIntervalMills = controllerFetchScheduleIntervalMills;
+    ScheduledExecutorService scheduledExecutorService = Executors.newSingleThreadScheduledExecutor();
+    scheduledExecutorService.scheduleAtFixedRate(() -> refreshBroker(), this.controllerFetchScheduleIntervalMills,
+        this.controllerFetchScheduleIntervalMills, TimeUnit.MILLISECONDS);
+    refreshBroker();
+  }
+
+  private void refreshBroker() {
+    Map<String, List<String>> tableToBrokerListMap = getTableToBrokersMap();
+    if (tableToBrokerListMap == null) {
+      throw new RuntimeException("Unable to fetch broker information from controller");
+    }
+    tableToBrokerListMapRef.set(ImmutableMap.copyOf(tableToBrokerListMap));
+    Set<String> brokerSet = new HashSet<>();
+    for (List<String> brokerList : tableToBrokerListMap.values()) {
+      brokerSet.addAll(brokerList);
+    }
+    allBrokerListRef.set(ImmutableList.copyOf(brokerSet));
+  }
+
+  private Map<String, List<String>> getTableToBrokersMap() {
+    int controllerIdx = RANDOM.nextInt(controllerUrls.length);
+    for (int i = 0; i < controllerFetchRetries; i++) {
+      try {
+        String brokerTableQueryResp = IOUtils.toString(
+            new URL(String.format(CONTROLLER_ONLINE_TABLE_BROKERS_MAP_URL_TEMPLATE, controllerUrls[controllerIdx])));
+        Map<String, List<String>> tablesToBrokersMap = OBJECT_MAPPER.readValue(brokerTableQueryResp, Map.class);
+        for (String table : tablesToBrokersMap.keySet()) {
+          List<String> brokerHostPorts = new ArrayList<>();
+          for (String broker : tablesToBrokersMap.get(table)) {
+            brokerHostPorts.add(broker.replace("Broker_", "").replace("_", ":"));
+          }
+          tablesToBrokersMap.put(table, brokerHostPorts);
+        }
+        return tablesToBrokersMap;
+
+      } catch (Exception e) {
+        LOGGER.warn(String.format("Unable to fetch controller broker information from '%s', retry in %d millseconds",
+            controllerUrls[controllerIdx], this.controllerFetchRetriesIntervalMills), e);
+        controllerIdx = (controllerIdx + 1) % controllerUrls.length;
+        try {
+          Thread.sleep(controllerFetchRetriesIntervalMills);
+        } catch (InterruptedException interruptedException) {
+          // Swallow
+        }
+      }
+    }
+    LOGGER.warn(
+        String.format("Failed to fetch controller broker information with %d retries", this.controllerFetchRetries));
+    return null;
+  }
+
+  @Override
+  public String selectBroker(String table) {
+    if (table == null) {
+      List<String> list = allBrokerListRef.get();
```

Comment:
`allBrokerListRef` being computed outside of this method makes the api vulnerable to getting a broker that is no longer ONLINE in EV.

Reply:
This is true, I'm caching the broker list with a scheduled call.  I don't want to let each query calling controller very time.